File: lib/cfr/builtins/libs/dart/ui/shadow.dart
import 'dart:core';
import 'dart:ui';

import 'package:hydro_sdk/cfr/runtimeSupport.dart';

class VMManagedShadow extends VMManagedBox<Shadow> {
  VMManagedShadow(
      {required this.table, required this.vmObject, required this.hydroState})
      : super(
          table: table,
          vmObject: vmObject,
          hydroState: hydroState,
        ) {
    table['color'] = maybeBoxObject<Color>(
        object: vmObject.color, hydroState: hydroState, table: HydroTable());
    table['offset'] = maybeBoxObject<Offset>(
        object: vmObject.offset, hydroState: hydroState, table: HydroTable());
    table['blurRadius'] = vmObject.blurRadius;
    table['getBlurSigma'] =
        makeLuaDartFunc(func: (List<dynamic> luaCallerArguments) {
      return [
        vmObject.blurSigma,
      ];
    });
    table['toPaint'] =
        makeLuaDartFunc(func: (List<dynamic> luaCallerArguments) {
      return [
        maybeBoxObject<Paint>(
            object: vmObject.toPaint(),
            hydroState: hydroState,
            table: HydroTable()),
      ];
    });
    table['scale'] = makeLuaDartFunc(func: (List<dynamic> luaCallerArguments) {
      return [
        maybeBoxObject<Shadow>(
            object: vmObject.scale(luaCallerArguments[1]?.toDouble()),
            hydroState: hydroState,
            table: HydroTable()),
      ];
    });
    table['getHashCode'] =
        makeLuaDartFunc(func: (List<dynamic> luaCallerArguments) {
      return [
        vmObject.hashCode,
      ];
    });
    table['toString'] =
        makeLuaDartFunc(func: (List<dynamic> luaCallerArguments) {
      return [
        vmObject.toString(),
      ];
    });
  }

  final HydroTable table;

  final HydroState hydroState;

  final Shadow vmObject;
}

class RTManagedShadow extends Shadow implements Box<Shadow> {
  RTManagedShadow(
      {required double blurRadius,
      required Color color,
      required Offset offset,
      required this.table,
      required this.hydroState})
      : super(blurRadius: blurRadius, color: color, offset: offset) {
    table['vmObject'] = vmObject;
    table['unwrap'] = makeLuaDartFunc(func: (List<dynamic> luaCallerArguments) {
      return [unwrap()];
    });
    table['color'] = maybeBoxObject<Color>(
        object: color, hydroState: hydroState, table: HydroTable());
    table['offset'] = maybeBoxObject<Offset>(
        object: offset, hydroState: hydroState, table: HydroTable());
    table['blurRadius'] = blurRadius;
    table['_dart_getBlurSigma'] =
        makeLuaDartFunc(func: (List<dynamic> luaCallerArguments) {
      return [super.blurSigma];
    });
    table['_dart_toPaint'] =
        makeLuaDartFunc(func: (List<dynamic> luaCallerArguments) {
      return [
        maybeBoxObject<Paint>(
            object: super.toPaint(),
            hydroState: hydroState,
            table: HydroTable())
      ];
    });
    table['_dart_scale'] =
        makeLuaDartFunc(func: (List<dynamic> luaCallerArguments) {
      return [
        maybeBoxObject<Shadow>(
            object: super.scale(luaCallerArguments[1]?.toDouble()),
            hydroState: hydroState,
            table: HydroTable())
      ];
    });
    table['_dart_getHashCode'] =
        makeLuaDartFunc(func: (List<dynamic> luaCallerArguments) {
      return [super.hashCode];
    });
    table['_dart_toString'] =
        makeLuaDartFunc(func: (List<dynamic> luaCallerArguments) {
      return [super.toString()];
    });
  }

  final HydroTable table;

  final HydroState hydroState;

  Shadow unwrap() => this;
  Shadow get vmObject => this;
  @override
  double get blurSigma {
    Closure closure = table["getBlurSigma"];
    return closure.dispatch([table], parentState: hydroState)[0]?.toDouble();
  }

  @override
  Paint toPaint() {
    Closure closure = table["toPaint"];
    return maybeUnBoxAndBuildArgument<Paint, dynamic>(
        closure.dispatch([table], parentState: hydroState)[0],
        parentState: hydroState);
  }

  @override
  Shadow scale(double factor) {
    Closure closure = table["scale"];
    return maybeUnBoxAndBuildArgument<Shadow, dynamic>(
        closure.dispatch([table], parentState: hydroState)[0],
        parentState: hydroState);
  }

  @override
  int get hashCode {
    Closure closure = table["getHashCode"];
    return closure.dispatch([table], parentState: hydroState)[0];
  }

  @override
  String toString() {
    Closure closure = table["__tostring"];
    return closure.dispatch([table], parentState: hydroState)[0];
  }
}

void loadShadow({required HydroState hydroState, required HydroTable table}) {
  table['shadow'] = makeLuaDartFunc(func: (List<dynamic> luaCallerArguments) {
    return [
      RTManagedShadow(
          table: luaCallerArguments[0],
          hydroState: hydroState,
          blurRadius: luaCallerArguments.length >= 2
              ? luaCallerArguments[1]['blurRadius']
              : null?.toDouble(),
          color: maybeUnBoxAndBuildArgument<Color, dynamic>(
              luaCallerArguments.length >= 2
                  ? luaCallerArguments[1]['color']
                  : null,
              parentState: hydroState),
          offset: maybeUnBoxAndBuildArgument<Offset, dynamic>(
              luaCallerArguments.length >= 2
                  ? luaCallerArguments[1]['offset']
                  : null,
              parentState: hydroState))
    ];
  });
  table['shadowConvertRadiusToSigma'] =
      makeLuaDartFunc(func: (List<dynamic> luaCallerArguments) {
    return [
      Shadow.convertRadiusToSigma(luaCallerArguments[1]?.toDouble()),
    ];
  });
  table['shadowLerp'] =
      makeLuaDartFunc(func: (List<dynamic> luaCallerArguments) {
    final returnValue = Shadow.lerp(
        maybeUnBoxAndBuildArgument<Shadow?, dynamic>(luaCallerArguments[1],
            parentState: hydroState),
        maybeUnBoxAndBuildArgument<Shadow?, dynamic>(luaCallerArguments[2],
            parentState: hydroState),
        luaCallerArguments[3]?.toDouble());
    if (returnValue != null) {
      return [
        maybeBoxObject<Shadow?>(
            object: returnValue, hydroState: hydroState, table: HydroTable()),
      ];
    }
    return [];
  });
  table['shadowLerpList'] =
      makeLuaDartFunc(func: (List<dynamic> luaCallerArguments) {
    final returnValue = Shadow.lerpList(
        maybeUnBoxAndBuildArgument<List<Shadow>?, Shadow>(luaCallerArguments[1],
            parentState: hydroState),
        maybeUnBoxAndBuildArgument<List<Shadow>?, Shadow>(luaCallerArguments[2],
            parentState: hydroState),
        luaCallerArguments[3]?.toDouble());
    if (returnValue != null) {
      return [
        maybeBoxObject<List<dynamic>>(
            object: returnValue
                .map((x) => maybeBoxObject<Shadow>(
                    object: x, hydroState: hydroState, table: HydroTable()))
                .toList(),
            hydroState: hydroState,
            table: HydroTable()),
      ];
    }
    return [];
  });
  registerBoxer<Shadow>(boxer: (
      {required Shadow vmObject,
      required HydroState hydroState,
      required HydroTable table}) {
    return VMManagedShadow(
        vmObject: vmObject, hydroState: hydroState, table: table);
  });
}

File: runtime/dart/ui/shadow.ts
import { IList } from "../core/list";
import { Color, IColor } from "./color";
import { IOffset, Offset } from "./offset";
import { IPaint } from "./paint";
declare const dart: {
    ui: {
        shadow: (
            this: void,
            shadow: IShadow,
            props: { blurRadius: number; color: IColor; offset: IOffset }
        ) => IShadow;
        shadowConvertRadiusToSigma: (radius: number) => number;
        shadowLerp: (
            a: IShadow | undefined,
            b: IShadow | undefined,
            t: number
        ) => IShadow | undefined;
        shadowLerpList: (
            a: IList<IShadow> | undefined,
            b: IList<IShadow> | undefined,
            t: number
        ) => IList<IShadow> | undefined;
    };
};
export interface IShadow {
    color: IColor;
    offset: IOffset;
    blurRadius: number;
    getBlurSigma: () => number;
    toPaint: () => IPaint;
    scale: (factor: number) => IShadow;
    getHashCode: () => number;
    toString: () => string;
}
export class Shadow {
    public static _kColorDefault = 0xff000000;
    public static _kBytesPerShadow = 16;
    public static _kColorOffset = 0 << 2;
    public static _kXOffset = 1 << 2;
    public static _kYOffset = 2 << 2;
    public static _kBlurOffset = 3 << 2;
    public readonly color: IColor = undefined as any;
    public readonly offset: IOffset = undefined as any;
    public readonly blurRadius: number = undefined as any;
    public constructor(props: {
        blurRadius?: number;
        color?: IColor;
        offset?: IOffset;
    }) {
        dart.ui.shadow(this, {
            ...shadowDefaultProps,
            ...props,
        });
    }
    public static convertRadiusToSigma(radius: number): number {
        return dart.ui.shadowConvertRadiusToSigma(radius);
    }
    public static lerp(
        a: IShadow | undefined,
        b: IShadow | undefined,
        t: number
    ): IShadow | undefined {
        return dart.ui.shadowLerp(a, b, t);
    }
    public static lerpList(
        a: IList<IShadow> | undefined,
        b: IList<IShadow> | undefined,
        t: number
    ): IList<IShadow> | undefined {
        return dart.ui.shadowLerpList(a, b, t);
    }
    private readonly _dart_getBlurSigma: () => number = undefined as any;
    private readonly _dart_toPaint: () => IPaint = undefined as any;
    private readonly _dart_scale: (factor: number) => IShadow =
        undefined as any;
    private readonly _dart_getHashCode: () => number = undefined as any;
    private readonly _dart_toString: () => string = undefined as any;
    public getBlurSigma(): number {
        return this._dart_getBlurSigma();
    }
    public toPaint(): IPaint {
        return this._dart_toPaint();
    }
    public scale(factor: number): IShadow {
        return this._dart_scale(factor);
    }
    public getHashCode(): number {
        return this._dart_getHashCode();
    }
    public toString(): string {
        return this._dart_toString();
    }
}
const shadowDefaultProps = {
    blurRadius: 0.0,
    color: new Color(Shadow._kColorDefault),
    offset: Offset.zero,
};
